core/src/main/java/jenkins/security/UserDetailsCache.java

```diff
@@ -0,0 +1,190 @@
+    public static UserDetailsCache get() {
+        return ExtensionList.lookup(UserDetailsCache.class).get(UserDetailsCache.class);
+    }
+
+    /**
+     * Gets the cached UserDetails for the given username.
+     * Similar to {@link #loadUserByUsername(String)} except it doesn't perform the actual lookup if there is a cache miss.
+     *
+     * @param idOrFullName the username
+     *
+     * @return {@code null} if the cache doesn't contain any data for the key or the user details cached for the key.
+     * @throws UsernameNotFoundException if a previous lookup resulted in the same
+     */
+    @CheckForNull
+    public UserDetails getCached(String idOrFullName) throws UsernameNotFoundException {
+        Boolean exists = existanceCache.getIfPresent(idOrFullName);
+        if (exists != null && !exists) {
+            throw new UserMayOrMayNotExistException(String.format("\"%s\" does not exist", idOrFullName));
+        } else {
+            return detailsCache.getIfPresent(idOrFullName);
+        }
+    }
+
+    /**
+     * Locates the user based on the username, by first looking in the cache and then delegate to
+     * {@link hudson.security.SecurityRealm#loadUserByUsername(String)}.
+     *
+     * @param idOrFullName the username
+     * @return the details
+     *
+     * @throws UsernameNotFoundException (normally a {@link hudson.security.UserMayOrMayNotExistException})
+     *              if the user could not be found or the user has no GrantedAuthority
+     * @throws DataAccessException if user could not be found for a repository-specific reason
+     * @throws ExecutionException if anything else went wrong in the cache lookup/retrieval
+     */
+    @Nonnull
+    public UserDetails loadUserByUsername(String idOrFullName) throws UsernameNotFoundException, DataAccessException, ExecutionException {
+        Boolean exists = existanceCache.getIfPresent(idOrFullName);
+        if(exists != null && !exists) {
+            throw new UsernameNotFoundException(String.format("\"%s\" does not exist", idOrFullName));
+        } else {
+            try {
+                return detailsCache.get(idOrFullName, new Retriever(idOrFullName));
+            } catch (ExecutionException | UncheckedExecutionException e) {
+                if (e.getCause() instanceof UsernameNotFoundException) {
+                    throw ((UsernameNotFoundException)e.getCause());
+                } else if (e.getCause() instanceof DataAccessException) {
+                    throw ((DataAccessException)e.getCause());
+                } else {
+                    throw e;
+                }
+            }
+        }
+    }
+
+    /**
+     * Discards all entries in the cache.
+     */
+    public void invalidateAll() {
+        existanceCache.invalidateAll();
+        detailsCache.invalidateAll();
+    }
+
+    /**
+     * Discards any cached value for key.
+     * @param idOrFullName the key
+     */
+    public void invalidate(final String idOrFullName) {
```

Comment:
Caveat: don't we need to invoke this when deleting users to remove their details from the cache?  Quite possible I'm missing something, but it feels like a 🐛  not to invalidate the cache upon FullName modification or user deletion.